rfcs/rfc-006-budget-contract-language.md

```diff
@@ -0,0 +1,117 @@
+
+### Transfer (aka, Budget v2) Transaction Format
+```rust
+pub enum TransferCondition {
+    /// Wait for a `Timestamp` `Instruction` at or after the given `DateTime` from `Pubkey`
+    Timestamp(DateTime<Utc>, Pubkey),
+
+    /// Wait for a `Witness` `Instruction` from `Pubkey`
+    Witness(Pubkey),
+}
+
+pub struct Transfer {
+    /// Amount to be transferred
+    tokens: i64,
+
+    /// `Pubkey` that the `Cancel` `Instruction` may be issued from prior to the
+    // `Transfer` meeting all `conditions`
+    cancellable: Maybe<Pubkey>,
+
+    /// The list of conditions that must be met before the `Transfer` is unlocked
+    conditions: Vec<TransferCondition>,
+}
+
+/// The `Transaction` `userdata` contain a serialized `TransferInstruction`
+///
+/// * key[0] - the `caller` key
+/// * key[1] - the contract key
+///
+pub enum TransferInstruction {
+    /// Declare and instantiate a new `Transfer` into `key[1]` using tokens from `key[0]`.
+    ///
+    /// `key[1]` must be unallocated before calling `Create`
+    Create(Transfer),
```

Comment:
This is something like `NewContract(Contract(Instruction(Budget)))` in current code. I like the idea of moving all Budget-specific constructors out of `Instruction`.